[Woo POS] Display store details from POS settings and customize footer text in POS emails block format

jaclync opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOOMOB-460

This pull request applies the same customizations to Point of Sale (POS) completed and refunded order emails in block format as in HTML/plaintext formats - displaying available store details from POS settings, and customizing the footer text with a refactoring on how the footer text is rendered.

### Changes to POS Email Templates:
* Added support for displaying store details (name, email, phone, address) and a refund/returns policy in POS completed and refunded order emails. These details are pulled from POS settings or default values. (`plugins/woocommerce/templates/emails/block/general-block-email.php`)
* Replaced the hardcoded footer text in email templates with a dynamic personalization tag (`woocommerce/footer-text`) that adapts based on the email context (e.g., POS store name or site title). (`plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTags/StoreTagsProvider.php`, `plugins/woocommerce/src/Internal/EmailEditor/EmailTemplates/WooEmailTemplate.php`) [[1]](diffhunk://#diff-7a034f779dc1fdffa77685dab9c3cf0001b6d670a46fd11a573a3f43ac68c39fR99-R117) [[2]](diffhunk://#diff-ea1d686151eb1b8749ab827c40837af54d5f7a9c9f222c3653af69fd50643ae1L58-R58)

### Personalization Tag Updates:
* Introduced a new personalization tag, `woocommerce/footer-text`, to dynamically generate footer content for POS-related emails. This tag is shown in the default template to replace the previous text. (`plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTags/StoreTagsProvider.php`)

### Code Simplification and Cleanup:
* Removed redundant and hardcoded content from the `customer-pos-completed-order.php` and `customer-pos-refunded-order.php` templates, consolidating logic into the generalized `general-block-email.php` file. (`plugins/woocommerce/templates/emails/block/customer-pos-completed-order.php`, `plugins/woocommerce/templates/emails/block/customer-pos-refunded-order.php`) [[1]](diffhunk://#diff-e434a774bc8eff8ec95a119cfcaf8a762b045ee360993166de748c5a0e7d5948L24-L56) [[2]](diffhunk://#diff-935221350bf7781284d0db9989b194c940430f6384eeee61efb5f4407befed4cL33-L66)

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

\ | before | after
-- | -- | --
POS completed order (above order summary) | <img width="562" alt="Screenshot 2025-05-19 at 3 23 08 PM" src="https://github.com/user-attachments/assets/499984a9-0bb8-4de5-b4e9-1e4d0d3a5bdf" /> | <img width="563" alt="Screenshot 2025-05-19 at 3 23 57 PM" src="https://github.com/user-attachments/assets/497fe8c5-9761-4632-b02b-21174fc7690e" />
POS completed order (below order summary) | <img width="559" alt="Screenshot 2025-05-19 at 3 23 27 PM" src="https://github.com/user-attachments/assets/45c398e8-748b-4610-80a9-7fa9f800e3ab" /> | <img width="561" alt="Screenshot 2025-05-19 at 3 24 21 PM" src="https://github.com/user-attachments/assets/30828f60-196f-490e-bf51-175d444a79ef" />
POS refunded order (above order summary) | <img width="563" alt="Screenshot 2025-05-19 at 3 32 11 PM" src="https://github.com/user-attachments/assets/fce3b41a-dfa5-47e6-a036-5ae77926c347" /> | <img width="566" alt="Screenshot 2025-05-19 at 3 27 34 PM" src="https://github.com/user-attachments/assets/c25dbbed-abd2-4ce9-b865-2d0627b8d32d" />
POS refunded order (below order summary) | <img width="562" alt="Screenshot 2025-05-19 at 3 32 24 PM" src="https://github.com/user-attachments/assets/cd2ccc05-0207-41b4-9670-353e0881f549" /> | <img width="560" alt="Screenshot 2025-05-19 at 3 27 45 PM" src="https://github.com/user-attachments/assets/c3caebaa-08ec-44a6-8a1c-7f5c95f20fb1" />

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Prerequisites: the store has a way to see email logs (like with [WP Mail Logging plugin](https://wordpress.org/plugins/wp-mail-logging/)) or send emails.

- Run PR branch in development environment (e.g. `pnpm run --filter="@woocommerce/plugin-woocommerce" watch:build`), or test with Woo Beta Tester plugin and enable the `point-of-sale` admin feature flag (example p1746734640520419/1746734508.133719-slack-C070SJRA8DP)
- Enable block email editor in WC settings > Advanced > Features
- Go to WC settings > Emails > POS completed order > edit --> notice that in the post, there's a `woocommerce/footer-text` tag at the bottom of the template
- Go to WC settings > Emails > POS refunded order > edit --> notice that in the post, there's a `woocommerce/footer-text` tag at the bottom of the template
- Go to WC settings > Point of Sale tab, note the values set here
- Make an API request to create a POS order ("created_via": "pos-rest-api") with a billing email and at least one line item for refund testing later --> the request should succeed with an order ID in the response

<details>
<summary>Click to see example request body</summary>

```json
{
    "status": "completed",
    "currency": "USD",
    "billing": {
        "first_name": "",
        "last_name": "",
        "company": "",
        "address_1": "",
        "address_2": "",
        "city": "",
        "state": "",
        "postcode": "",
        "country": "",
        "email": "test@email.co",
        "phone": ""
    },
    "meta_data": [
        {
            "key": "_wc_order_attribution_source_type",
            "id": 0,
            "value": "mobile_app"
        },
        {
            "key": "_cash_change_amount",
            "value": "0"
        },
        {
            "key": "_charge_id",
            "value": "auth2j9wq8"
        }
    ],
    "line_items": [
        {
            "id": 0,
            "product_id": 40,
            "quantity": 2
        }
    ],
    "created_via": "pos-rest-api"
}
```

</details>

- With the ID of the created order, make a request to send POS completed order email `POST /wc/v3/orders/$orderId/actions/send_email` with request body `{"template_id": "customer_pos_completed_order"}` --> it should succeed with a message like "Email template &quot;POS completed order&quot; sent to ...". In email logs or if the store can send email, check the email that was just sent. Below the order summary, there should be a section with heading that shows the store name from POS settings, and email/phone/address below when available in POS settings. If refund & returns policy is set in POS settings, there should be a section below the store details with a heading and the policy text 
- For the same order ID, issue a refund in the app or core --> In email logs or if the store can send email, check the email that was just sent. The email should mention "partial refund" or "refund" in the subject. Below the order summary, there should be a section with heading that shows the store name from POS settings, and email/phone/address below when available in POS settings. If refund & returns policy is set in POS settings, there should be a section below the store details with a heading and the policy text
- Feel free to repeat the steps above with different POS settings

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

- [x] @jaclync tested that refund emails for non-POS orders remain the same
- [x] @jaclync tested partial and full refund emails for POS orders
- [x] @jaclync tested that customer invoice emails remain the same

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

jaclync committed:
Create POS block email tags provider and register them in POS completed order email class.

jaclync committed:
Replace personalization tags with conditional content in the general block for POS emails because personalization tag blocks cannot be hidden when the content is empty.

jaclync committed:
Remove unnecessary content in POS refund email block template that is moved to general-block-email.

jaclync committed:
Revert "Create POS block email tags provider and register them in POS completed order email class."

This reverts commit f6fd40a52659f91e8ec31c4877178f09e5a44f9f.

jaclync committed:
Show placeholder heading when store name is empty.

jaclync committed:
Move default template footer text to a store personalization tag that shows different text based on email class ID for POS email customizations.

jaclync committed:
Update copy to be consistent with HTML/plaintext templates.

jaclync committed:
Add changelog.

jaclync committed:
Bump template WC versions.

triple0t commented on plugins/woocommerce/templates/emails/block/general-block-email.php:
```diff
@@ -116,3 +141,59 @@
+			<h3><?php echo esc_html( get_option( 'woocommerce_pos_store_name', PointOfSaleDefaultSettings::get_default_store_name() ) ); ?></h3>
+			<!-- /wp:heading -->
+			<?php else : ?>
+			<!-- wp:heading {"level":3} -->
```
Please add class `class="wp-block-heading"` to the headings to prevent this error `Updated Block: core/heading` from logging in the console

jaclync replied:
Good call, added in https://github.com/woocommerce/woocommerce/pull/58100/commits/2ed61d9df65a472858aaf839d72b27afbd12c9a7.

triple0t commented on plugins/woocommerce/templates/emails/block/general-block-email.php:
```diff
@@ -82,6 +83,30 @@
 	<?php
 endif;
 
+if ( 'customer_pos_completed_order' === $email->id || 'customer_pos_refunded_order' === $email->id ) :
+	?>
+	<!-- wp:paragraph -->
+	<p>
+	<?php
+	if ( ! empty( $order->get_billing_first_name() ) ) {
+		/* translators: %s: Customer first name */
+		printf( esc_html__( 'Hi %s,', 'woocommerce' ), esc_html( $order->get_billing_first_name() ) );
+	} else {
+		echo esc_html__( 'Hi there,', 'woocommerce' );
+	}
+	?>
+	</p>
+	<!-- /wp:paragraph -->
+	<!-- wp:paragraph -->
+	<p>
+	<?php
+		echo esc_html__( 'Here’s a reminder of what you’ve bought:', 'woocommerce' );
+	?>
+	</p>
+	<!-- /wp:paragraph -->
+	<?php
+endif;
```
I'm not convinced this block of text should be here.
I understand the use case for the section below, but this one can actually live in the block email templates.

As much as possible, we don't want to put too much custom content here as this is a work-in-progress file we are using for now, pending when we turn the whole Woo Orders content into a block.

You can achieve the same purpose by updating the customer first-name tag
e.g Adding `return isset( $parameters['default'] ) ? $parameters['default'] : '';` to [here](https://github.com/woocommerce/woocommerce/blob/ba4d2a7e703532b9871402f400d5bcfef6e2f757/plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTags/CustomerTagsProvider.php#L48).

and using 
```html
<!-- wp:paragraph -->
<p>
<?php
	/* translators: %s: Customer first name */
	printf( esc_html__( 'Hi %s,', 'woocommerce' ), '<!--[woocommerce/customer-first-name default="there"]-->' );
?>
</p>
<!-- /wp:paragraph -->
```
in the block HTML template.

jaclync replied:
The reason for not using the empty default value is because there will be an empty space between "Hi" and the comma, like `Hi ,` which feels not so ideal. Especially, for the upcoming POS milestone p91TBi-cGi-p2, adding a customer isn't included and most POS orders will have no customers and thus this fallback case.

> I'm not convinced this block of text should be here.
I understand the use case for the section below, but this one can actually live in the block email templates.
> As much as possible, we don't want to put too much custom content here as this is a work-in-progress file we are using for now, pending when we turn the whole Woo Orders content into a block.

I understand the rationale here, thanks for sharing the context. Given that customers aren't currently supported in POS, I will remove this conditional content here and just show `Hi there,` for the POS emails in the block templates.

triple0t replied:
> The reason for not using the empty default value is because there will be an empty space between "Hi" and the comma, like Hi , which feels not so ideal. Especially, for the upcoming POS milestone p91TBi-cGi-p2, adding a customer isn't included and most POS orders will have no customers and thus this fallback case.

Yes, I understand. This is why I suggested [this](https://github.com/woocommerce/woocommerce/pull/58100#discussion_r2101770808) workaround

Instead of having `Hi ,`

You would instead be having `Hi DEFAULT_VALUE,` which in this case is `Hi there,`

Directly inline with your implementation [here](https://github.com/woocommerce/woocommerce/blob/b2f3fb10962fc264068a08402520972664113c7d/plugins/woocommerce/templates/emails/block/general-block-email.php#L95)

jaclync replied:
How about localization, do translators know how to translate "there" in this specific context? In mobile, we generally avoid combining localizable substrings into one string because it might be challenging to localize the substrings individually that are meant for one context.

I think just displaying "Hi there" would be the best solution until we add customers support in POS. Updated in https://github.com/woocommerce/woocommerce/pull/58100/commits/c5984e6b51e6a0689b8e40815e47a960a32ed909.

triple0t commented on plugins/woocommerce/src/Internal/EmailEditor/PersonalizationTags/StoreTagsProvider.php:
```diff
@@ -95,5 +96,24 @@ function ( array $context ): string {
+				__( 'Footer text', 'woocommerce' ),
+				'woocommerce/footer-text',
+				__( 'Store', 'woocommerce' ),
+				function ( array $context ): string {
+					if ( isset( $context['wc_email'] ) && ( 'customer_pos_completed_order' === $context['wc_email']->id || 'customer_pos_refunded_order' === $context['wc_email']->id ) ) {
+						$pos_store_name = htmlspecialchars_decode( get_option( 'woocommerce_pos_store_name', PointOfSaleDefaultSettings::get_default_store_name() ) );
+						/* translators: %s: name of the Point of Sale store */
+						return sprintf( __( 'You received this email because you shopped at %s', 'woocommerce' ), $pos_store_name );
+					}
+
+					$site_title = htmlspecialchars_decode( get_bloginfo( 'name' ) );
+					/* translators: %s: site title */
+					return sprintf( __( 'You received this email because you shopped on %s', 'woocommerce' ), $site_title );
+				},
```
Question:
Instead of registering a new Footer text personalization tag, why not move this logic to the Site-title tag?

It should work in there since we want to show a different site title for POS emails and a different site title for other emails.

jaclync replied:
It was more because of the "shopped on" phrase, which is more for the e-commerce context (shopping on a website) and less applicable to in-person purchases. Is it possible to update the footer copy from "shopped on" to "shopped at" for all core emails? I think "shopped at" is appropriate for both online and in-person sales, would like to hear your thoughts. If so, I will remove this footer tag, update the copy and update the site-title tag to return the POS store name for POS orders.

If you think changing the footer copy is too risky, I can also revert the footer changes and leave it as a future improvement.

triple0t replied:
> Is it possible to update the footer copy from "shopped on" to "shopped at" for all core emails?

Yes. It's possible.

Please feel free to update it.

>  I think "shopped at" is appropriate for both online and in-person sales, would like to hear your thoughts.

I agree with you. You can change it.

jaclync replied:
Reverted the previous footer & tag changes in https://github.com/woocommerce/woocommerce/pull/58100/commits/90e9f171e4b901b6ecc236f3be4d956ee676fcf4, and updated the footer copy and site-title value for POS orders in https://github.com/woocommerce/woocommerce/pull/58100/commits/49bfc1fdf3be7942b71cd8ab4e156c618a1e80a2.

triple0t commented on plugins/woocommerce/templates/emails/block/general-block-email.php:
```diff
@@ -100,7 +100,7 @@
 	<p>
 	<?php
-		echo esc_html__( 'Here’s a reminder of what you’ve ordered:', 'woocommerce' );
+		echo esc_html__( 'Here’s a reminder of what you’ve bought:', 'woocommerce' );
```
May I ask why you do not want this section to be in the block email template?

jaclync replied:
It was because this paragraph is displayed below the previously conditional greeting content and above the order summary that are both in the custom block. Now that we no longer need the conditional greeting in https://github.com/woocommerce/woocommerce/commit/c5984e6b51e6a0689b8e40815e47a960a32ed909, this section is now in the block templates.

triple0t reviewed (commented):
Thanks for the PR.

I dropped some comments and asked for additional feedback regarding testing on Slack.

PS: Can you replace `Shoppers` with `Customer` for the emails ([WC_Email_Customer_POS_Refunded_Order](https://github.com/woocommerce/woocommerce/blob/ba4d2a7e703532b9871402f400d5bcfef6e2f757/plugins/woocommerce/includes/emails/class-wc-email-customer-pos-refunded-order.php#L64), [WC_Email_Customer_POS_Completed_Order](https://github.com/woocommerce/woocommerce/blob/ba4d2a7e703532b9871402f400d5bcfef6e2f757/plugins/woocommerce/includes/emails/class-wc-email-customer-pos-completed-order.php#L52))?  --> p1747812776646409/1747132360.600409-slack-C040TSG288Y.

jaclync committed:
Merge branch 'trunk' into feat/WOOMOB-460-pos-email-block-format-v2

jaclync committed:
Replace conditional content in POS emails above order summary with static content in block templates as customers are not supported in POS.

jaclync committed:
Add `wp-block-heading` class to heading tag.

jaclync committed:
Fix incorrect POS phone option name.

jaclync committed:
Revert "Move default template footer text to a store personalization tag that shows different text based on email class ID for POS email customizations."

This reverts commit 5f4deb2df9215b6404abc9cf1cb1613d1331d2a0.

jaclync committed:
Support POS store name and POS-friendly copy in the footer of block email template.

jaclync committed:
Polish static text in block templates.

jaclync commented:
Thanks for the feedback @triple0t! I updated the PR and responded to your latest comments. Tomorrow I will do more testing and see if I can add some unit tests. I will also prepare a test site for easier testing. Thanks for trying different things to test the API before.

triple0t reviewed (approved):
LGTM. Thank you for implementing this.

### Testing result.

**With Block Email Editor Active**

| Name | Image |
| ------ | ----- |
|   POS completed order    |   <img width="957" alt="Screenshot 2025-05-22 at 3 00 34 PM" src="https://github.com/user-attachments/assets/64d44b8e-c423-4906-88a2-3e011434a5c5" />    |
| Order refunded |  <img width="742" alt="Screenshot 2025-05-22 at 3 01 32 PM" src="https://github.com/user-attachments/assets/a192ca0a-bfbe-4528-8787-84a49f00281e" /> | 



**With Block Email Editor deactivated**

| Name | Image |
| ------ | ----- |
|    POS completed order    |   <img width="670" alt="Screenshot 2025-05-22 at 3 03 26 PM" src="https://github.com/user-attachments/assets/5bafc568-619f-4091-9956-9e5841ba6e19" />    |
| Order refunded |  <img width="585" alt="Screenshot 2025-05-22 at 3 04 54 PM" src="https://github.com/user-attachments/assets/99698d3a-49bd-4c5e-9343-33121f85a946" /> | 


**Without Custom POS options**

<img width="1049" alt="Screenshot 2025-05-22 at 3 13 57 PM" src="https://github.com/user-attachments/assets/f1f113dc-0dd0-4175-9b33-f82dd9ebdf6a" />

<img width="877" alt="Screenshot 2025-05-22 at 3 15 24 PM" src="https://github.com/user-attachments/assets/f1f08d66-8ff6-4c8b-8772-64bf0733da12" />

triple0t commented:
@jaclync, Please fix the [lint](https://github.com/woocommerce/woocommerce/actions/runs/15188133816/job/42714088014?pr=58100) error and you can merge.
Thank you

jaclync committed:
Replace "shoppers" copy with "customers" re p1747812776646409/1747132360.600409-slack-C040TSG288Y.

jaclync committed:
Handle empty POS store name better in email content and block template footer.

jaclync committed:
Try fixing lint error on missing flags parameter for `htmlspecialchars_decode`.

jaclync commented:
Thanks a bunch for testing with the screenshots & the feedback! I updated the PR and did another round of testing.

---

> Please fix the [lint](https://github.com/woocommerce/woocommerce/actions/runs/15188133816/job/42714088014?pr=58100) error and you can merge.

Fixed in https://github.com/woocommerce/woocommerce/pull/58100/commits/a6a7bb4f96d7589ef4fa818d4d84859489d4382d.

> > PS: Can you replace Shoppers with Customer for the emails ([WC_Email_Customer_POS_Refunded_Order](https://github.com/woocommerce/woocommerce/blob/ba4d2a7e703532b9871402f400d5bcfef6e2f757/plugins/woocommerce/includes/emails/class-wc-email-customer-pos-refunded-order.php#L64), [WC_Email_Customer_POS_Completed_Order](https://github.com/woocommerce/woocommerce/blob/ba4d2a7e703532b9871402f400d5bcfef6e2f757/plugins/woocommerce/includes/emails/class-wc-email-customer-pos-completed-order.php#L52))? --> p1747812776646409/1747132360.600409-slack-C040TSG288Y.
> 
> Kind reminder to update this section.

Thanks for the reminder! Updated in https://github.com/woocommerce/woocommerce/pull/58100/commits/b09aca004e288714afdb861ddc58f46ba2b49a2a.

> I think it would also be nice to account for cases where these options are blank.

Good suggestion. The store name seems to look the most buggy in the email subject and footer, when the value is set to empty in the POS settings because `wp_option` default value is only used when the option value is not set. In https://github.com/woocommerce/woocommerce/pull/58100/commits/7c9c9ef9b460d8bb4791f23c5dba99b863799645, I updated the site title/name value to return the default value (WP site title) if the POS store name is empty for POS emails.

jaclync committed:
Merge branch 'trunk' into feat/WOOMOB-460-pos-email-block-format-v2

jaclync commented:
@triple0t I plan to merge this when CI passes after merging trunk, please still feel free to share any other suggestions if you see anything in the latest commits. I can address them if before the merge, or separately.

triple0t commented:
Looks good to me. Thank you, @jaclync.
